Edit usaco-596 with faster solution (#1828)
cpinitiative/usaco-guide: solutions/usaco-596.mdx

id: usaco-596
source: USACO Silver 2016 January
title: Build Gates

> authors: Maggie Liu, Aadit Ambadkar

Reviewer: ```suggestion
author: Maggie Liu, Aadit Ambadkar
```
needs to be like this to work

Author: Done

	}
}
```
## Alternate Faster Solution

> 

Reviewer 2: add time complexity here as well

Author: Done

Reviewer: @nwatx shouldn't time complexity go under the explanation like [here](https://usaco.guide/general/adding-solution?lang=cpp#steps)?

It is possible to make the following 2 observations:

1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate.

> 2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, if FJ walks along an edge from point a to b, and he did not walk from a to b in the past, and he did not walk from b to a in the past, and b is a point he had already visited, then he has created exactly 1 new region.

Reviewer 2: this is a run on sentence. Maybe consider breaking it as a bulleted list and saying these are the conditions that must be satisfied.

Author: Done

> Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster, and works in O(nlog(n)) due to log(n) set lookup and insertion. 

Reviewer 2: format these complexities with latex

Author: Done

	}
}
```

> ## Alternate Faster Solution - Time Complexity $\mathcal{O}(N\log N)$

Reviewer 2: ![image](https://user-images.githubusercontent.com/43394822/123534040-a89f5f00-d6df-11eb-922b-0a128d6fa993.png)
usually how its done^

Author: Done

It is possible to make the following 2 observations:

1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate.

> 2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous close region if:

Reviewer 2: maybe clarify something like "if all the following conditions are satisfied"

Author: Done

Author: Done

	* FJ walks along an edge from point a to b
	* He did not walk from a to b in the past
	* He did not walk from b to a in the past

> 	* b is a point he had already visited

Reviewer 2: latex variables

Author: Done

Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster, and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion. 

> ```cpp

Reviewer: use tabs instead of spaces to indent

Author: Done (I think)

#include <bits/stdc++.h>
using namespace::std;

int main() {
    freopen("gates.in", "r", stdin);
    freopen("gates.out", "w", stdout);
	
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n; cin >> n;
    string s; cin >> s;
	
    set<pair<pair<int, int>, pair<int, int>>> visedge;
    set<pair<int, int>> visnode;
    int ans = 0;
    pair<int, int> prev = {0, 0};
    visnode.insert(prev);
	
    for (int i = 0; i < n; i++) {
        int x = prev.first; int y = prev.second;

>         if (s[i]=='N') {

Reviewer: ```suggestion
        if (s[i] == 'N') {
```
can you add spaces to all the operators below too?

Author: Done

            x++;
        } else if (s[i]=='S') {
            x--;
        } else if (s[i]=='E') {
            y--;
        } else {
            y++;
        }
			  
        if (visedge.find({{x, y}, prev})==visedge.end() && visnode.find({x, y})!=visnode.end()) {
            ans++;
        }
			  
        visedge.insert({{x, y}, prev});
        visedge.insert({prev, {x, y}});
        visnode.insert({x, y});
			  
        prev.first = x;
        prev.second = y;
    }
			  
    cout << ans << "\n";
}
```

> </CPPSection>

Reviewer: right now, we have
```
<LanguageSection>
<CPPSection>
first sol
second sol
</CPPSection>
</LanguageSection>
```
separate the two solutions into two different languagesections like
```
<LanguageSection>
<CPPSection>
first sol
</CPPSection>
</LanguageSection>
<LanguageSection>
<CPPSection>
second sol
</CPPSection>
</LanguageSection>
```

Author: Done

</CPPSection>

</LanguageSection>

<LanguageSection>

<CPPSection>

> ## Alternate Faster Solution

Reviewer: can you format it like [this](https://usaco.guide/general/adding-solution?lang=cpp#steps)? move the explanation and time complexity outside of the languagesection and put the time complexity below the explanation

Author: Done

Reviewer: > put the time complexity below the explanation

Author: Oh ok.

Author: Done.

1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate.

> 2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous close region if all of these conditions are met:

Reviewer: ```suggestion
2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous closed region if all of these conditions are met:
```

It is possible to make the following 2 observations:

1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate.
2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous close region if all of these conditions are met:
	* FJ walks along an edge from point $a$ to $b$
	* He did not walk from $a$ to $b$ in the past
	* He did not walk from $b$ to $a$ in the past
	* $b$ is a point he had already visited

> Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster, and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion. 

Reviewer 2: this shouldn't be in the cpp section, move to outside language section

Author: Done

</CPPSection>

</LanguageSection>
	
## Alternate Faster Solution

It is possible to make the following 2 observations:

1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate.
2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous closed region if all of these conditions are met:
	* FJ walks along an edge from point $a$ to $b$
	* He did not walk from $a$ to $b$ in the past
	* He did not walk from $b$ to $a$ in the past
	* $b$ is a point he had already visited

Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster, and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion. 

> **Time Complexity:** $\mathcal{O}(N\log N)$

Reviewer 2: move to line 102

Reviewer: oops i think i commented on the wrong comment but https://github.com/cpinitiative/usaco-guide/pull/1828#discussion_r659398168

Author: @maggie-j-liu told me to move it down there

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace::std;

int main() {
	freopen("gates.in", "r", stdin);
	freopen("gates.out", "w", stdout);
	
	ios::sync_with_stdio(0);
	cin.tie(0);

	int n; cin >> n;
	string s; cin >> s;
	
	set<pair<pair<int, int>, pair<int, int>>> visedge;
	set<pair<int, int>> visnode;
	int ans = 0;
	pair<int, int> prev = {0, 0};
	visnode.insert(prev);
	
	for (int i = 0; i < n; i++) {
		int x = prev.first; int y = prev.second;
		if (s[i] == 'N') {
			x++;
		} else if (s[i] == 'S') {
			x--;
		} else if (s[i] == 'E') {
			y--;
		} else {
			y++;
		}
			  

> 		if (visedge.find({{x, y}, prev}) == visedge.end() && visnode.find({x, y}) != visnode.end()) {

Reviewer 2: past the 80 character limit

Author: What should I do?

Reviewer: can you put the `&&` and part after on a new line?

Author: Ok, Did that

</CPPSection>

</LanguageSection>
	
## Alternate Faster Solution

It is possible to make the following 2 observations:

1. For each contiguous closed region, there must be 1 gate. In a minimum state, each region has exactly 1 gate.
2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous closed region if all of these conditions are met:
	* FJ walks along an edge from point $a$ to $b$
	* He did not walk from $a$ to $b$ in the past
	* He did not walk from $b$ to $a$ in the past
	* $b$ is a point he had already visited

> Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster, and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion.

Reviewer 2: ```suggestion
Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion.
```

> Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is almost 10 times faster, and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion.

Reviewer 2: would be helpful to mention relative to what solution (10 times faster)

Author: How does it look now?

Author: @nwatx

> Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is about 10 times faster than using flood fill and works in $\mathcal{O}(N\log N)$ due to $\log N$ set lookup and insertion.
> 	
> **Time Complexity:** $\mathcal{O}(N\log N)$

Reviewer 3: ```suggestion
Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is about 10 times faster than flood filling due to $\log N$ lookup and insertion.
	
## Implementation

**Time Complexity:** $\mathcal{O}(N\log N)$
```

2. Each time FJ comes across a node he has visited along an edge he has not crossed in either direction, he creates exactly 1 new region. In other words, FJ creates exactly 1 new contiguous closed region if all of these conditions are met:
	* FJ walks along an edge from point $a$ to $b$
	* He did not walk from $a$ to $b$ in the past
	* He did not walk from $b$ to $a$ in the past
	* $b$ is a point he had already visited

Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is about 10 times faster than flood filling due to $\log N$ lookup and insertion.
	
## Implementation

**Time Complexity:** $\mathcal{O}(N\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("gates.in", "r", stdin);
	freopen("gates.out", "w", stdout);
	
	ios::sync_with_stdio(0);
	cin.tie(0);

	int n; cin >> n;
	string s; cin >> s;
	
	set<pair<pair<int, int>, pair<int, int>>> visedge;
	set<pair<int, int>> visnode;
	int ans = 0;

> 	pair<int, int> prev = {0, 0};

Reviewer 4: ```suggestion
	pair<int, int> prev{0, 0};
```
more concise

Author: The equals is more clear

Reviewer: not sure how it's more clear since they mean the exact same thing??

Reviewer 4: not really? think list initialization is fairly standard?

	visnode.insert(prev);
	
	for (int i = 0; i < n; i++) {
		int x = prev.first; int y = prev.second;

> 		if (s[i] == 'N') {

Reviewer 4: switch statement maybe

Author: No, you need to check each time.

Reviewer: i think sans meant to use a https://en.cppreference.com/w/cpp/language/switch

Reviewer 4: ? it's a buch of if/elifs not a bunch of ifs?

Author: I didn't understand what you mean, so there was a bit of confusion. Sorry. I don't think a switch is needed.

Reviewer 4: isn't needed, but it's best practice

Reviewer 5: I don't think we have a preference on this. Either is fine... (actually I prefer if statements lol)

Author: Great! Can you merge this if it is done?

			x++;
		} else if (s[i] == 'S') {
			x--;
		} else if (s[i] == 'E') {
			y--;
		} else {
			y++;
		}
			  

> 		if (visedge.find({{x, y}, prev}) == visedge.end() 

Reviewer 4: maybe a small comment or two because i have no clue what this is supposed to do

Author: That is standard C++ set find... Its how you check if an element does not exist in a set.

Reviewer 4: of course i know what it does code-wise
i mean like what doing this accomplishes
"check if ...."

Author: Ohhh ok

Author: Done

Reviewer 4: remove line 155?

Author: Ok, makes sense

				&& visnode.find({x, y}) != visnode.end()) {
			ans++;
		}
			  
		visedge.insert({{x, y}, prev});
		visedge.insert({prev, {x, y}});
        	visnode.insert({x, y});
			  
		prev.first = x;
		prev.second = y;
	}

> 			  

Reviewer 4: ```suggestion
```

Author: Those lines are needed though. Either X or Y change.

Reviewer: the suggestion is to delete the blank line (line 166) (not sure what you see but github diffs can be weird sometimes)

Reviewer 4: what

	* FJ walks along an edge from point $a$ to $b$
	* He did not walk from $a$ to $b$ in the past
	* He did not walk from $b$ to $a$ in the past
	* $b$ is a point he had already visited

Under these observations, we can use a set of visited edges and visited nodes, and keep track of the number of contiguous closed regions. This solution is about 10 times faster than flood filling due to $\log N$ lookup and insertion.
	
## Implementation

**Time Complexity:** $\mathcal{O}(N\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("gates.in", "r", stdin);
	freopen("gates.out", "w", stdout);
	
	ios::sync_with_stdio(0);
	cin.tie(0);

	int n; cin >> n;
	string s; cin >> s;
	
	set<pair<pair<int, int>, pair<int, int>>> visedge;
	set<pair<int, int>> visnode;
	int ans = 0;
	pair<int, int> prev{0, 0};
	visnode.insert(prev);
	
	for (int i = 0; i < n; i++) {
		int x = prev.first; int y = prev.second;
		if (s[i] == 'N') {
			x++;
		} else if (s[i] == 'S') {
			x--;
		} else if (s[i] == 'E') {
			y--;
		} else {
			y++;
		}
			      

> 		// The below checks if the edge from prev to x,y was inside the visited edges. 
> 		// See the 2nd observation for more information   

Reviewer 4: ```suggestion
		// Check if the edge from prev to {x, y} was visited (see 2nd observation)
```
